README.md
**ledger-wallet /**
[Readme](https://cosmic.plus/#view:js-ledger-wallet)
• [Examples](https://cosmic.plus/#view:js-ledger-wallet/EXAMPLES)
• [Contributing](https://cosmic.plus/#view:js-ledger-wallet/CONTRIBUTING)
• [Changelog](https://cosmic.plus/#view:js-ledger-wallet/CHANGELOG)

# Readme

![Licence](https://img.shields.io/github/license/cosmic-plus/js-ledger-wallet.svg)
[![Dependencies](https://badgen.net/david/dep/cosmic-plus/js-ledger-wallet)](https://david-dm.org/cosmic-plus/js-ledger-wallet)
![Vulnerabilities](https://snyk.io/test/npm/@cosmic-plus/ledger-wallet/badge.svg)
![Bundle](https://badgen.net/badgesize/gzip/cosmic-plus/js-ledger-wallet-web/master/ledger-wallet.js?label=bundle)
![Downloads](https://badgen.net/npm/dt/@cosmic-plus/ledger-wallet)

Easy Ledger Wallet support for Stellar applications and command line.

(Weekly updates: [Reddit](https://reddit.com/r/cosmic_plus),
[Twitter](https://twitter.com/cosmic_plus),
[Keybase](https://keybase.io/team/cosmic_plus),
[Telegram](https://t.me/cosmic_plus))

## Introduction

This library is a convenient wrapper around the official Ledger libraries for
Stellar:

- [Stellar app API](https://www.npmjs.com/package/@ledgerhq/hw-app-str)
- [Transport Node HID](https://www.npmjs.com/package/@ledgerhq/hw-transport-node-hid) - Node.js support
- [Transport U2F](https://www.npmjs.com/package/@ledgerhq/hw-transport-u2f) - Browser support

It provides a way to support Ledger Wallets with a few one-liners:

```js
// Step 1: Connect
await ledgerWallet.connect()

// Step 2: Get public key
const pubkey = ledgerWallet.publicKey

// Step 3: Sign transaction
await ledgerWallet.sign(transaction)

// Extra: Event handlers
ledgerWallet.onConnect = connectionHandler
ledgerWallet.onDisconnect = disconnectionHandler
```

This library is compatible with both Node.js and browser environments.

## Installation

### NPM/Yarn

- NPM: `npm install @cosmic-plus/ledger-wallet`
- Yarn: `yarn add @cosmic-plus/ledger-wallet`

In your script: `const ledgerWallet = require("@cosmic-plus/ledger-wallet")`

### Bower

`bower install cosmic-plus-ledger-wallet`

In your HTML page:

```HTML
<script src="./bower_components/cosmic-plus-ledger-wallet/ledger-wallet.js"></script>
```

### CDN

In your HTML page:

```HTML
<script src="https://cdn.cosmic.plus/ledger-wallet@2.x"></script>
```

_Note:_ For production release it is advised to serve your copy of the library.

## Usage

### Functions

#### await ledgerWallet.connect([account])

Waits for a connection with the Ledger Wallet application for Stellar. If
**account** is not provided, account 1 is used. The library will stop
listening for a connection if `await ledgerWallet.disconnect()` is called.

Once the connection is established, you can use `await ledgerWallet.connect(account)` again at any time to ensure the device is
still connected.

When switching to another account, you can `await ledgerWallet.connect(new_account)` without prior disconnection.

_Note:_ To stay consistent with the way Trezor numbers accounts, **account**
starts at 1 (derivation path: `m/44'/148'/0'`).

| Param     | Type                 | Default | Description                                                                         |
| --------- | -------------------- | ------- | ----------------------------------------------------------------------------------- |
| [account] | `Number` \| `String` | `1`     | Either an account number (starts at 1) or a derivation path (e.g: `m/44'/148'/0'`). |

#### await ledgerWallet.sign(transaction)

Prompts the user to accept **transaction** using the connected account of
their Ledger Wallet.

If the user accepts, adds the signature to **transaction**. Else, throws an
error.

| Param       | Type          | Description              |
| ----------- | ------------- | ------------------------ |
| transaction | `Transaction` | A StellarSdk Transaction |

#### await ledgerWallet.disconnect()

Close the connection with the Ledger device, or stop waiting for one in case
a connection has not been established yet.

#### ledgerWallet.newAccount([horizon])

Connects the first unused account.

_Note:_ merged accounts are considered as used.

| Param     | Type                 | Default                                   | Description                                                                                                       |
| --------- | -------------------- | ----------------------------------------- | ----------------------------------------------------------------------------------------------------------------- |
| [horizon] | `String` \| `Server` | `&quot;https://horizon.stellar.org&quot;` | The Horizon server where to check for account existence. It can be either an URL or a _StellarSdk.Server_ object. |

#### ledgerWallet.scan([params]) ⇒ `Array`

Scans the ledger device for accounts that exist on **params.horizon**. The
scanning stops after encountering **params.attempts** unused accounts.

Merged accounts are considered as existing accounts and will reset the
**params.attempts** counter when encountered.

Returns an _Array_ of _Objects_ containing `account` number, `publicKey`,
`path`, and `state` (either `"open"` or `"merged"`).

| Param                  | Type                 | Default                                   | Description                                                                                                       |
| ---------------------- | -------------------- | ----------------------------------------- | ----------------------------------------------------------------------------------------------------------------- |
| [params]               | `Object`             |                                           | Optional parameters.                                                                                              |
| [params.horizon]       | `String` \| `Server` | `&quot;https://horizon.stellar.org&quot;` | The Horizon server where to check for account existence. It can be either an URL or a _StellarSdk.Server_ object. |
| [params.attempts]      | `Number`             | `3`                                       | The number of empty accounts before scanning stops.                                                               |
| [params.includeMerged] | `Boolean`            | `false`                                   | List merged accounts as well.                                                                                     |

#### ledgerWallet.getPublicKeys([start], [length]) ⇒ `Array`

Request multiple public keys from the Ledger device. The request will return
**length** accounts, starting by **start** (minimum 1).

Returns an _Array_ of _Objects_ with properties `account`, `publicKey`, and
`path`.

| Param    | Type     | Default | Description                     |
| -------- | -------- | ------- | ------------------------------- |
| [start]  | `Number` | `1`     | Starting account number         |
| [length] | `Number` | `1`     | Number of account to be listed. |

### Events

#### ledgerWallet.onConnect : `function`

_Function_ to execute on each connection.

#### ledgerWallet.onDisconnect : `function`

_Function_ to execute on each disconnection.

### Members

#### ledgerWallet.publicKey : `String`

Public key of the connected account.

#### ledgerWallet.path : `String`

Derivation path of the connected account (default: `m/44'/148'/0'`).

#### ledgerWallet.version : `String`

Version of the Stellar application installed on the connected device.

#### ledgerWallet.multiOpsEnabled : `Boolean`

Whether or not the user accepts to sign transactions without checking them on
the device first. This allows to sign long transactions (10+ ops) that could
normally not be handled by the device memory, but this is insecure.

#### ledgerWallet.transport : `Transport`

The Ledger Transport instance. (internal component)

#### ledgerWallet.application : `StellarApp`

The Ledger Stellar application instance. (internal component)

## Links

**Organization:** [Cosmic.plus](https://cosmic.plus/) | [@GitHub](https://git.cosmic.plus) | [@NPM](https://www.npmjs.com/search?q=cosmic-plus)

**Follow:** [Reddit](https://reddit.com/r/cosmic_plus) | [Twitter](https://twitter.com/cosmic_plus) | [Medium](https://medium.com/cosmic-plus) | [Codepen](https://codepen.io/cosmic-plus)

**Talk:** [Telegram](https://t.me/cosmic_plus) | [Keybase](https://keybase.io/team/cosmic_plus)


CHANGELOG.md
**ledger-wallet /**
[Readme](https://cosmic.plus/#view:js-ledger-wallet)
• [Examples](https://cosmic.plus/#view:js-ledger-wallet/EXAMPLES)
• [Contributing](https://cosmic.plus/#view:js-ledger-wallet/CONTRIBUTING)
• [Changelog](https://cosmic.plus/#view:js-ledger-wallet/CHANGELOG)

# Changelog

All notable changes to this project will be documented in this file.

This project adheres to **[Semantic
Versioning](https://semver.org/spec/v2.0.0.html)**. Version syntax is
`{major}.{minor}.{patch}`, where a field bump means:

- **Patch**: The release contains bug fixes.
- **Minor**: The release contains backward-compatible changes.
- **Major**: The release contains compatibility-breaking changes.

**Remember:** Both micro and minor releases are guaranteed to respect
backward-compatibility and can be updated to without risk of breakage. For major
releases, please check this changelog before upgrading.

## 2.8.0 - 2021-09-17

### Changed

- Meta: Update dependencies.
- Meta: Update @legderhq libraries to 6.x.

## 2.7.0 - 2021-01-08

### Changed

- Logic: Update Ledger libraries to 5.38.x.

## 2.6.0 - 2020-11-08

### Changed

- Logic: Update Ledger libraries to 5.28.x.

## 2.5.0 - 2020-09-12

### Changed

- Logic: Update @ledger libraries to 5.23.x.

## 2.4.0 - 2020-06-21

### Changed

- Meta: Switch to `@kisbox/helpers`.

### Fixed

- Meta: Remove redundant dependency.

## 2.3.0 - 2020-04-18

### Changed

- Logic: Update [@ledgerhq] libraries. (HID transport updates)

## 2.2.0 - 2020-02-01

### Changed

- Logic: Update @ledger libraries to 5.7.x. (dependencies updates, better
  errors)

## 2.1.0 - 2020-01-04

### Changed

- Logic: Upgrade @ledgerhq libraries to 5.x. This saves 16KiB or more when the
  library gets bundled.
- Logic: Improve device waiting. This should prevent errors due to device not
  being ready yet without re-introducing the invite spamming issue. If there's
  still issues in specific environments, please let me know on
  Keybase#cosmic_plus.

## 2.0.1 - 2019-12-29

### Chore

- Documentation: update README.md.

## 2.0.0 - 2019-12-28

### Breaking

- API: `.connect()` always connect to account 1. To ensure the device is still
  connected, code must now explicitly pass the same account parameter than for
  initialization (e.g: `.connect(4)`). Before `.connect()` was enough but it led
  to unexpected behavior.
- API: Account numbering now starts at 1. This is to be consistent with
  Ledger/Trezor account numbering. Code using `.connect(number)` must be updated
  accordingly. `.connect()` continues to select `m/44'/148'/0'` as before.
- API: `.path` is now properly prefixed by `m/`. (e.g.: `m/44'/148'/0'` instead
  of `44'/148'/0'`)

### Removed

- API: REMOVE deprecated properties & parameters. The following `ledgerWallet`
  properties have been removed: `.account`, `.index`, `.internalFlag`, and
  `.error`. The related parameters have been removed from `.connect()`
  (parameters 2 and 3 are now ignored).

### Added

- API: Add `.newAccount()`. This method connects the first unused account of a
  Ledger device. ([See
  documentation](https://cosmic.plus/#view:js-ledger-wallet/%23ledgerwalletnewaccount))
- API: Add `.scan()`. This method scans for existing accounts on the Ledger
  device. ([See
  documentation](https://cosmic.plus/#view:js-ledger-wallet/%23ledgerwalletscan))
- API: Add `.getPublicKeys()`. This method retrieves multiple public keys at
  once. ([See
  documentation](https://cosmic.plus/#view:js-ledger-wallet/%23ledgerwalletgetpublickeys))

### Fixed

- API: Ensure `.path` and `publicKey` are set only when connected.
- Logic: Fix the timing of `connect()`/`disconnect()`.

## 1.5.0 - 2019-11-26

### Changed

- Logic: Remove device polling. @cosmic-plus/ledger-wallet included logic for
  detecting device disconnections. However, this logic was causing alert
  spamming in some environments. Because there's no way to probe the device
  without making those messages appear, the only sane solution was to remove
  this feature.

### Fixed

- Logic: Fix `.connect()` errors handling. The old logic was causing some
  browsers to spam device requests.

## 1.4.0 - 2019-11-23

### Changed

- Logic: Update [@ledgerhq] libraries to 4.74.2. (bugfixes)

## 1.3.1 - 2019-11-09

### Changed

- Documentation: Update documentation.

## 1.3.0 - 2019-11-02

### Deprecated

- API: Deprecate `index` & `internalFlag` .connect() parameters. Those
  parameters were used to construct derivation paths after [BIP44]
  specifications. However, those are rarely used in Stellar & were removed from
  this library documentation months ago.

  The new way to specify arbitrary derivation paths is to pass them explicitly
  (e.g.: "m/44'/148'/1'/2'").

### Added

- API: Add support for arbitrary derivation paths. Example:
  `ledger.connect("m/44'/255'/5'")`

### Fixed

- Logic: Always use `null` for empty variables. The code was not consistent in
  its usage of `null` and `undefined`.

## 1.2.0 - 2019-10-26

### Changed

- Logic: Update Ledger libraries to 4.73.x. (bugfixes)

### Fixed

- Logic: Fix two timing bugs. - `onConnect` callback was fired before the end
  of initialization. - In some cases, polling could cause an alive session to
  disconnect.

## 1.1.2 - 2019-09-14

### Fixed

- Meta: Remove stellar-sdk from bower dependencies.

## 1.1.1 - 2019-08-31

### Fixed

- Documentation: Fix contributing guidelines.

## 1.1.0 - 2019-08-17

### Added

- Doc: Add EXAMPLES.md.

## 1.0.0 - 2019-08-10

**Note:** This major release doesn't contain any breaking change. The version
bump only means that this library is public and stable.

### Fixed

- Make bundlers pick the transpiled code.

## 0.3.6 - 2019-07-31

### Changed

- Remove `ledgerWallet.connect()` dependency on StellarSdk.

## 0.3.5 - 2019-07-26

### Changed

- Automate release procedure.
- Add contributing guidelines.
- Update Ledger libraries to 4.68.x

## 0.3.4 - 2019-07-20

### Changed


test/signing.js
"use_strict"
/**
 * Signing test
 */

const StellarSdk = require("stellar-sdk")
const { misc } = require("@cosmic-plus/jsutils")

const ledger = require("../src")

// eslint-disable-next-line no-console
console.log(ledger)

StellarSdk.Network.useTestNetwork()

async function test () {
  await ledger.connect()

  const pubkey = ledger.publicKey
  // eslint-disable-next-line no-console
  console.log(`Public Key: ${pubkey}`)
  const tx = makeTransaction(pubkey)

  try {
    // eslint-disable-next-line no-console
    console.log("Waiting for confirmation...")
    await ledger.sign(tx)
    // eslint-disable-next-line no-console
    console.log("Transaction signed")
    await ledger.disconnect()
  } catch (error) {
    console.error(error)
    // eslint-disable-next-line no-console
    console.log("Retry in 5 seconds\n")
    misc.timeout(5000).then(test)
  }
}

function makeTransaction (pubkey) {
  const account = new StellarSdk.Account(pubkey, "0")
  const txBuilder = new StellarSdk.TransactionBuilder(account, { fee: 100 })
  return txBuilder.setTimeout(0).build()
}

test()


test/connection.js
"use_strict"
/**
 * Connection/disconnection test
 */

const { misc } = require("@cosmic-plus/jsutils")
const ledger = require("../src")

// eslint-disable-next-line no-console
console.log(ledger)

ledger.onConnect = function () {
  // eslint-disable-next-line no-console
  console.log(ledger.path + ": " + ledger.publicKey)
}

ledger.onDisconnect = function () {
  misc.timeout(1000).then(test)
}

async function test () {
  await ledger.connect(0)
}

test()


test/index.spec.js
/* eslint-env jasmine */
"use strict"

const StellarSdk = require("stellar-sdk")
const { friendbot } = require("@cosmic-plus/base")

const ledgerWallet = require("../src")

const { any } = jasmine

/* Setup */

const horizon = "https://horizon-testnet.stellar.org"
const xdr =
  "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAZAAAAAAAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAUAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=="

const tx1 = new StellarSdk.Transaction(xdr, StellarSdk.Networks.TESTNET)
const tx2 = new StellarSdk.Transaction(xdr, StellarSdk.Networks.TESTNET)

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000

/* Specifications */

describe("ledgerWallet", () => {
  afterEach(async () => {
    ledgerWallet.onConnect = undefined
    ledgerWallet.onDisconnect = undefined
    await ledgerWallet.disconnect()
  })

  describe("connect()", () => {
    it("connects with Ledger Wallet", async () => {
      await ledgerWallet.connect()
    })

    it("sets .path using default", async () => {
      await ledgerWallet.connect()
      expect(ledgerWallet.path).toMatch(/^m\/44'\/148'\/0'$/)
    })

    it("sets .path using number", async () => {
      await ledgerWallet.connect(4)
      expect(ledgerWallet.path).toMatch(/^m\/44'\/148'\/3'$/)
    })

    it("sets .path using custom derivation path (1)", async () => {
      await ledgerWallet.connect("m/44'/148'/2'")
      expect(ledgerWallet.path).toMatch(/^m\/44'\/148'\/2'$/)
    })

    it("sets .path using custom derivation path (2)", async () => {
      await ledgerWallet.connect("m/44'/148'/4'/1'/6'")
      expect(ledgerWallet.path).toMatch(/^m\/44'\/148'\/4'\/1'\/6'$/)
    })

    it("sets .publicKey", async () => {
      await ledgerWallet.connect()
      expect(ledgerWallet.publicKey).toEqual(any(String))
      expect(ledgerWallet.publicKey.length).toEqual(56)
      StellarSdk.Keypair.fromPublicKey(ledgerWallet.publicKey)
    })

    it("sets properties related to stellar-app-ledger", async () => {
      await ledgerWallet.connect()
      expect(ledgerWallet.version).toEqual(any(String))
      expect(ledgerWallet.multiOpsEnabled).toEqual(any(Boolean))
    })

    it("sets misc properties", async () => {
      await ledgerWallet.connect()
      expect(ledgerWallet.transport).toEqual(any(Object))
      expect(ledgerWallet.application).toEqual(any(Object))
    })
  })

  describe("sign()", () => {
    it("signs transactions", async () => {
      await ledgerWallet.connect()
      // eslint-disable-next-line no-console
      console.log("Please accept the dummy transaction")
      await ledgerWallet.sign(tx1)
      const hasSigned = StellarSdk.Utils.verifyTxSignedBy
      const pubkey = ledgerWallet.publicKey
      expect(hasSigned(tx1, pubkey)).toBe(true)
    })

    it("throws when user refuses to sign", async () => {
      await ledgerWallet.connect()
      // eslint-disable-next-line no-console
      console.log("Please reject the dummy transaction")
      return ledgerWallet
        .sign(tx2)
        .then(() => {
          return new Error("Did not throw an error")
        })
        .catch((error) => expect(error).toEqual(any(Error)))
    })
  })

  describe(".disconnect()", () => {
    it("disconnects from Ledger Wallet", async () => {
      await ledgerWallet.connect()
      await ledgerWallet.disconnect()
    })

    it("unsets session properties", async () => {
      await ledgerWallet.connect()
      await ledgerWallet.disconnect()
      expect(ledgerWallet.path).toBe(null)
      expect(ledgerWallet.publicKey).toBe(null)
      expect(ledgerWallet.version).toEqual(null)
      expect(ledgerWallet.multiOpsEnabled).toEqual(null)
      expect(ledgerWallet.transport).toEqual(null)
      expect(ledgerWallet.application).toEqual(null)
    })
  })

  describe(".scan()", () => {
    beforeAll(async () => {
      // eslint-disable-next-line no-console
      console.log("Ensure that test accounts exist...")
      await wantAccount(1)
      await wantAccount(2)
      await wantAccount(5)
      await wantAccount(10)
    })

    it("scans for accounts (retry: 0)", async () => {
      const accounts = await ledgerWallet.scan({ horizon, attempts: 1 })
      expect(accounts).toEqual(testAccounts.slice(0, 2))
    })

    it("scans for accounts (retry: default = 3)", async () => {
      const accounts = await ledgerWallet.scan({ horizon })
      expect(accounts).toEqual(testAccounts.slice(0, 3))
    })

    it("scans for accounts (retry: 5)", async () => {
      const accounts = await ledgerWallet.scan({ horizon, attempts: 5 })
      expect(accounts).toEqual(testAccounts)
    })
  })

  describe(".newAccount()", () => {
    it("connect the next unused account", async () => {
      const accounts = await ledgerWallet.scan({
        horizon,
        attempts: 1,
        includeMerged: true
      })
      const last = accounts[accounts.length - 1]
      await ledgerWallet.newAccount(horizon)

      const accountIndex = +ledgerWallet.path.replace(/^.*([0-9]+)'/, "$1")
      if (last) expect(accountIndex).toEqual(last.account)
      else expect(accountIndex).toEqual(0)
    })
  })

  describe("events", () => {
    it("calls .onConnect on connection", async () => {
      let connect = false
      ledgerWallet.onConnect = () => connect = true
      await ledgerWallet.connect()
      expect(connect).toBe(true)
    })

    it("calls .onDisconnect on disconnection", async () => {
      let disconnect = false
      ledgerWallet.onDisconnect = () => disconnect = true
      await ledgerWallet.connect()
      await ledgerWallet.disconnect()
      expect(disconnect).toBe(true)
    })
  })
})

/* Helpers */
let testAccounts = []

async function wantAccount (account) {
  await ledgerWallet.connect(account)
  await friendbot(ledgerWallet.publicKey).catch(() => null)

  testAccounts.push({
    account,
    publicKey: ledgerWallet.publicKey,
    state: "open",
    path: ledgerWallet.path
  })

  await ledgerWallet.disconnect()
}


src/index.js
"use strict"
/**
 * This library is a convenient wrapper around the official Ledger libraries for
 * Stellar:
 *
 * - [Stellar app API](https://www.npmjs.com/package/@ledgerhq/hw-app-str)
 * - [Transport Node HID](https://www.npmjs.com/package/@ledgerhq/hw-transport-node-hid) - Node.js support
 * - [Transport U2F](https://www.npmjs.com/package/@ledgerhq/hw-transport-u2f) - Browser support
 *
 * It provides a way to support Ledger Wallets with a few one-liners:
 *
 * ```js
 * // Step 1: Connect
 * await ledgerWallet.connect()
 *
 * // Step 2: Get public key
 * const pubkey = ledgerWallet.publicKey
 *
 * // Step 3: Sign transaction
 * await ledgerWallet.sign(transaction)
 *
 * // Extra: Event handlers
 * ledgerWallet.onConnect = connectionHandler
 * ledgerWallet.onDisconnect = disconnectionHandler
 * ```
 *
 * This library is compatible with both Node.js and browser environments.
 *
 * @exports ledgerWallet
 */
const ledger = exports

const { isNode, nodeRequire, timeout } = require("@kisbox/helpers")

const StellarApp = require("@ledgerhq/hw-app-str").default
const Transport = isNode
  ? nodeRequire("@ledgerhq/hw-transport-node-hid").default
  : require("@ledgerhq/hw-transport-u2f").default

/* Configuration */
const BIP32_PATH = "m/44'/148'"

/* Properties */

/**
 * Public key of the connected account.
 * @var {String}
 */
ledger.publicKey = null

/**
 * Derivation path of the connected account (default: `m/44'/148'/0'`).
 * @var {String}
 */
ledger.path = null

/**
 * Version of the Stellar application installed on the connected device.
 * @var {String}
 */
ledger.version = null

/**
 * Whether or not the user accepts to sign transactions without checking them on
 * the device first. This allows to sign long transactions (10+ ops) that could
 * normally not be handled by the device memory, but this is insecure.
 * @var {Boolean}
 */
ledger.multiOpsEnabled = null

/**
 * The Ledger Transport instance. (internal component)
 * @var {Transport}
 */
ledger.transport = null

/**
 * The Ledger Stellar application instance. (internal component)
 * @var {StellarApp}
 */
ledger.application = null

/* Methods */

let connection, disconnection

/**
 * Waits for a connection with the Ledger Wallet application for Stellar. If
 * **account** is not provided, account 1 is used. The library will stop
 * listening for a connection if `await ledgerWallet.disconnect()` is called.
 *
 * Once the connection is established, you can use `await
 * ledgerWallet.connect(account)` again at any time to ensure the device is
 * still connected.
 *
 * When switching to another account, you can `await
 * ledgerWallet.connect(new_account)` without prior disconnection.
 *
 * _Note:_ To stay consistent with the way Trezor numbers accounts, **account**
 * starts at 1 (derivation path: `m/44'/148'/0'`).
 *
 * @async
 * @param {Number|String} [account=1] - Either an account number (starts at 1)
 * or a derivation path (e.g: `m/44'/148'/0'`).
 */
ledger.connect = async function (account = 1) {
  const path = ledger.connect.path(account)
  if (disconnection) await disconnection
  if (ledger.path && ledger.path !== path) softReset()
  if (!connection) connection = connect(path)
  return connection
}

ledger.connect.path = function (account) {
  if (typeof account === "number") {
    if (account < 1) throw new Error("Account number starts at 1.")
    return `${BIP32_PATH}/${account - 1}'`
  } else {
    return account
  }
}

async function connect (path) {
  // eslint-disable-next-line no-console
  console.log("Attempting ledger connection...")
  connection = true

  // Try to connect until disconnect() is called or until connection happens.
  let startTime
  while (connection && !ledger.publicKey) {
    startTime = +new Date()

    try {
      if (!ledger.transport || isNode) {
        ledger.transport = await Transport.create()
      }
      if (!ledger.application || isNode) {
        ledger.application = new StellarApp(ledger.transport)
      }

      // Set ledger properties
      const app = ledger.application
      Object.assign(ledger, await app.getPublicKey(path.replace(/^m\//, "")))
      Object.assign(ledger, await app.getAppConfiguration())
      ledger.path = path

      onConnect()
    } catch (error) {
      if (error.id === "U2FNotSupported") {
        // This frame may show up when using strict Content-Security-Policy
        // See: https://github.com/LedgerHQ/ledgerjs/issues/254
        const iframeSelector = "iframe[src^=chrome-extension/*/u2f-comms.html]"
        const iframe = document.querySelector(iframeSelector)
        if (iframe) iframe.parentNode.removeChild(iframe)
      }

      // Keep looping over timeout/non-connected/non-opened app errors.
      const ellapsed = +new Date() - startTime
      if (
        // Firefox user cancels.
        error.id === "U2F_1"
        // Chrome/Win user cancels.
        || error.id === "U2F_4" && ellapsed < 25000
      ) {
        softReset()
        error.message = "Connexion invite cancelled"
        throw error
      } else if (ellapsed < 8000) {
        // Prevent blocking message loops in potentially unhandled cases.
        await timeout(8000 - ellapsed)
      }
    }
  }
}

/**
 * Prompts the user to accept **transaction** using the connected account of
 * their Ledger Wallet.
 *
 * If the user accepts, adds the signature to **transaction**. Else, throws an
 * error.
 *
 * @async
 * @param {Transaction} transaction - A StellarSdk Transaction
 */
ledger.sign = async function (transaction) {
  if (!ledger.publicKey) throw new Error("No ledger wallet connected.")

  const result = await ledger.application.signTransaction(
    ledger.path.replace(/^m\//, ""),
    transaction.signatureBase()
  )
  const signature = result.signature.toString("base64")
  transaction.addSignature(ledger.publicKey, signature)

  return transaction
}

/**
 * Close the connection with the Ledger device, or stop waiting for one in case
 * a connection has not been established yet.
 *
 * @async
 */
ledger.disconnect = async function () {
  const transport = ledger.transport
  reset()

  if (transport) {
    disconnection = closeTransport(transport)
    disconnection.then(onDisconnect)
    await disconnection
  } else if (disconnection) {
    await disconnection
  }
  disconnection = null
}

async function closeTransport (transport) {
  // If transport is not valid anymore we consider the transport as closed.
  try {
    await transport.close()
  } catch (error) {
    console.error(error)
  }
}

function softReset () {
  connection = null
  ledger.path = null
  ledger.publicKey = null
}

function reset () {
  connection = null
  libValues.forEach((key) => ledger[key] = null)
}

const libValues = [
  "transport",
  "application",
  "path",
  "version",
  "publicKey",
  "multiOpsEnabled"
]

/**
 * Connects the first unused account.
 *
 * _Note:_ merged accounts are considered as used.
 *
 * @param {String|Server} [horizon=https://horizon.stellar.org] - The
 * Horizon server where to check for account existence. It can be either an URL
 * or a _StellarSdk.Server_ object.
 */
ledger.newAccount = async function (horizon = "https://horizon.stellar.org") {
  const accounts = await ledger.scan({
    horizon,
    attempts: 1,
    includeMerged: true
  })
  const last = accounts[accounts.length - 1]
  await ledger.connect(last ? last.account + 1 : 1)
}

/**
 * Scans the ledger device for accounts that exist on **params.horizon**. The
 * scanning stops after encountering **params.attempts** unused accounts.
 *
 * Merged accounts are considered as existing accounts and will reset the
 * **params.attempts** counter when encountered.
 *
 * Returns an _Array_ of _Objects_ containing `account` number, `publicKey`,
 * `path`, and `state` (either `"open"` or `"merged"`).
 *
 * @param {Object} [params] - Optional parameters.
 * @param {String|Server} [params.horizon=https://horizon.stellar.org] - The
 * Horizon server where to check for account existence. It can be either an URL
 * or a _StellarSdk.Server_ object.
 * @param {Number} [params.attempts=3] - The number of empty accounts before
 * scanning stops.
 * @param {Boolean} [params.includeMerged=false] - List merged accounts as well.
 * @returns {Array}
 */
ledger.scan = async function ({
  horizon = "https://horizon.stellar.org",
  attempts = 3,
  includeMerged = false
}) {
  if (typeof horizon === "string") {
    const { Server } = require("@cosmic-plus/base/es5/stellar-sdk")
    horizon = new Server(horizon)
  }

  const accounts = []
  let miss = 0,
    index = 1

  while (miss < attempts) {
    const current = (await ledger.getPublicKeys(index, 1))[0]

    // effectForAccount let us know about merged account as well.
    const callBuilder = horizon.effects().forAccount(current.publicKey)
    const response = await callBuilder
      .order("desc")
      .limit(1)
      .call()
      .catch(() => null)

    if (response) {
      miss = 0
      const latest = response.records[0]
      current.state = latest.type === "account_removed" ? "merged" : "open"

      if (includeMerged || current.state === "open") {
        accounts.push(current)
      }
    } else {
      miss++
    }

    index++
  }

  return accounts
}

/**
 * Request multiple public keys from the Ledger device. The request will return
 * **length** accounts, starting by **start** (minimum 1).
 *
 * Returns an _Array_ of _Objects_ with properties `account`, `publicKey`, and
 * `path`.
 *
 * @param {Number} [start=1] - Starting account number
 * @param {Number} [length=1] - Number of account to be listed.
 * @return {Array}
 */
ledger.getPublicKeys = async function (start = 1, length = 1) {
  const keys = []

  for (let account = start; account < start + length; account++) {
    await ledger.connect(account)
    keys.push({
      account,
      publicKey: ledger.publicKey,
      path: ledger.path
    })
  }

  return keys
}

/* Events */

/**
 * _Function_ to execute on each connection.
 *
 * @category event
 * @var {Function}
 */
ledger.onConnect = null
async function onConnect () {
  // eslint-disable-next-line no-console
  console.log("Ledger connected")
  if (typeof ledger.onConnect === "function") ledger.onConnect()
}

/**
 * _Function_ to execute on each disconnection.
 *
 * @category event
 * @var {Function}
 */
ledger.onDisconnect = null
function onDisconnect () {
  // eslint-disable-next-line no-console
  console.log("Ledger disconnected")
  if (typeof ledger.onDisconnect === "function") ledger.onDisconnect()
}
